File: src/com/basmilius/ps/bastools/util/EditorUtils.java
package com.basmilius.ps.bastools.util;

import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.editor.*;
import com.intellij.openapi.project.Project;

public final class EditorUtils
{

	/**
	 * Inserts or replaces text in the editor.
	 *
	 * @param editor  Editor instance.
	 * @param project Project instance.
	 * @param str     String you want to add/replace.
	 *
	 * @throws Exception Thrown if inserting fails.
	 */
	public static void insertOrReplaceMultiCaret (final Editor editor, final Project project, final String str) throws Exception
	{
		final WriteCommandAction wca = new WriteCommandAction.Simple(project)
		{

			@Override
			protected final void run () throws Throwable
			{
				final Document document = editor.getDocument();
				final CaretModel caret = editor.getCaretModel();
				final SelectionModel selection = editor.getSelectionModel();

				if (selection.hasSelection())
				{
					final int[] starts = selection.getBlockSelectionStarts();
					final int[] ends = selection.getBlockSelectionEnds();
					final int length = starts.length - 1;

					for (int i = length; i >= 0; i--)
					{
						document.replaceString(starts[i], ends[i], str);
					}
				}
				else if (caret.getCaretCount() > 0)
				{
					for (final Caret c : caret.getAllCarets())
					{
						document.insertString(c.getOffset(), str);
						c.moveToOffset(c.getOffset() + 7);
					}
				}
				else
				{
					throw new Exception("No carets found!");
				}
			}

		};
		wca.execute();
	}

}

File: src/com/basmilius/ps/bastools/util/EditorUtils.kt
package com.basmilius.ps.bastools.util

import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.editor.*
import com.intellij.openapi.fileEditor.ex.FileEditorManagerEx
import com.intellij.openapi.fileEditor.impl.EditorWindow
import com.intellij.openapi.project.Project

/**
 * Object EditorUtils
 *
 * @author Bas Milius
 * @package com.basmilius.ps.bastools.util
 */
object EditorUtils
{

	/**
	 * Inserts or replaces text in the editor.
	 *
	 * @param editor  Editor instance.
	 * @param project Project instance.
	 * @param str     String you want to add/replace.
	 *
	 * @author Bas Milius
	 */
	@Throws(Exception::class)
	fun insertOrReplaceMultiCaret(editor : Editor, project : Project, str : String)
	{
		val wca = object : WriteCommandAction.Simple<Unit>(project)
		{

			@Throws(Throwable::class)
			override fun run()
			{
				val document = editor.document
				val caret = editor.caretModel
				val selection = editor.selectionModel

				if (selection.hasSelection())
				{
					val starts = selection.blockSelectionStarts
					val ends = selection.blockSelectionEnds
					val length = starts.size - 1

					for (i in length downTo 0)
					{
						document.replaceString(starts[i], ends[i], str)
					}
				}
				else if (caret.caretCount > 0)
				{
					for (c in caret.allCarets)
					{
						document.insertString(c.offset, str)
						c.moveToOffset(c.offset + 7)
					}
				}
				else
				{
					throw Exception("No carets found!")
				}
			}

		}
		wca.execute()
	}

	/**
	 * Switches to a specific tab.
	 *
	 * @param project Project
	 * @param context DataContext
	 * @param index Int
	 */
	fun switchToTab(project : Project, context : DataContext, index : Int)
	{
		val editorManager = FileEditorManagerEx.getInstanceEx(project)
		val currentWindow = EditorWindow.DATA_KEY.getData(context) ?: editorManager.currentWindow
		val files = currentWindow.files
		val fileIndex = if (index == -1) (files.size - 1) else index

		if (fileIndex >= 0 && fileIndex < files.size)
		{
			editorManager.openFile(files[fileIndex], true)
		}
	}

}
